content/posts/monad-expressions-in-csharp.md

```diff
@@ -0,0 +1,131 @@
++++
+title =  "C# expressions that try to be monadic"
+description =  ""
+date =  2021-07-01
+template = "blog/page.html"
+draft =  false
+
+[taxonomies]
+authors = ["Robert Marciniak"]
+#tags =  ["monad", "linq", "async"]
+#categories =  []
++++
+
+# Why we can't have nice things, c# query expression edition
+
+The closest thing to f#'s computational expression is either combination of `async` `await` keywords in contexts that
+do not always translate to asynchronous context or query expression, which does not always translate to querying something.
+
+That's because C# is a corporate enterprise language which tries to smuggle Nice Things through, but in a different,
+more corporate-enterprise-like form - not too abstract, sprinkled with "query SQL from code!"-, "Improve responsiveness without big refactors!"-, or similar seasoning.
+
+## 'what's the underlying type?'
+
+### Case 1
+
+```cs
+from index in Enumerable.Range(1, 5)
+from date in Enumerable.Repeat(DateTime.Today, 5)
+select (index, date) into pair
+select pair.date.AddDays(pair.index);
+```
+
+```fs
+monad {
+    let! index = Enumerable.Range(1, 5)
+    let! date = Enumerable.Repeat(DateTime.Today, 5)
+    let pair = (index, date)
+    yield pair |> (fun (i, d) -> d.AddDays(i |> float))
+}
+```
+
+### Case 2
+
+```cs
+from index in GetIndexAsync()
+from date in Task.FromResult(DateTime.Today)
+select (index, date) into pair
+select pair.date.AddDays(pair.index);
+```
+
+```fs
+monad {
+    let! index = GetIndexAsync()
+    let! date = Task.FromResult(DateTime.Today)
+    let pair = (index, date)
+    yield pair |> (fun (i, d) -> d.AddDays(i |> float))
+}
+```
+
+### Case 3
+```cs
+from index in Option.Some(5)

```

Comment:
Sadly, I don't believe it's possible, because `nullable` is opaque by design, you can't know whether it's `Nullable<T>` where `T` is copy type, or if it's a compiler trick to allow for `?` suffix for reference types.

Other than that, it would still need an introduction of `Nullable` static type that would hold `map`, `bind` and others, but you can't, because you cannot define generic as `Function<T?>()`